src/kit_version.rs

```diff
@@ -0,0 +1,207 @@
+use lazy_static::lazy_static;
+use regex::Regex;
+use serde::de::{self, Deserialize, Deserializer, Visitor};
+use serde::ser::{Serialize, Serializer};
+use std::convert::TryFrom;
+use std::fmt;
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct BuildInfo {
+    pub commit: String,
+    pub branch: Option<String>,
+}
+
+impl fmt::Display for BuildInfo {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self.branch.as_ref() {
+            None => write!(f, "{}", self.commit),
+            Some(branch) => write!(f, "{}@{}", self.commit, branch),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct KitVersion {
+    pub epoch: u16,
+    pub major: u8,
+    pub minor: u8,
+    pub patch: u8,
+    pub dev: bool,
+    pub build_info: Option<BuildInfo>,
+}
+
+impl fmt::Display for KitVersion {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.dev {
+            match &self.build_info {
+                Some(build_info) => write!(
+                    f,
+                    "{}.{}.{}.{}dev:{}",
+                    self.epoch, self.major, self.minor, self.patch, build_info
+                ),
+                None => write!(
+                    f,
+                    "{}.{}.{}.{}dev",
+                    self.epoch, self.major, self.minor, self.patch
+                ),
+            }
+        } else {
+            match &self.build_info {
+                Some(build_info) => write!(
+                    f,
+                    "{}.{}.{}.{}:{}",
+                    self.epoch, self.major, self.minor, self.patch, build_info
+                ),
+                None => write!(
+                    f,
+                    "{}.{}.{}.{}",
+                    self.epoch, self.major, self.minor, self.patch
+                ),
+            }
+        }
+    }
+}
+
+impl TryFrom<&str> for KitVersion {
+    type Error = &'static str;
+
+    fn try_from(val: &str) -> Result<Self, Self::Error> {
+        lazy_static! {
+            static ref RE: Regex = Regex::new(
+                "^(\\d+)\\.(\\d+)\\.(\\d+)\\.(\\d+)(dev)?(?::([0-9a-f]{5,40})(?:@(\\w+))?)?$"
+            )
+            .unwrap();
+        }
+        match RE.captures(val) {
+            Some(captures) => {
+                let epoch = captures.get(1).unwrap().as_str().parse::<u16>();
```

Comment:
Issue: The use of `unwrap` here will result in a panic if there's something completely malformed. I recommend [`Result::and_then`](https://doc.rust-lang.org/std/result/enum.Result.html#method.and_then)

eg `captures.get(1).and_then(c => c.as_str().parse::<u16>())` will be `Result<u16, E>`